jkube-kit/build/api/src/main/java/org/eclipse/jkube/kit/build/api/helper/BuildArgResolverUtil.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2019 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at:
+ *
+ *     https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.jkube.kit.build.api.helper;
+
+import com.google.common.collect.ImmutableMap;
+import org.apache.commons.lang3.StringUtils;
+import org.eclipse.jkube.kit.common.JKubeConfiguration;
+import org.eclipse.jkube.kit.config.image.ImageConfiguration;
+import org.eclipse.jkube.kit.config.image.build.BuildConfiguration;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+
+public class BuildArgResolverUtil {
+  private static final String ARG_PREFIX = "docker.buildArg.";
+
+  private BuildArgResolverUtil() { }
+
+  public static Map<String, String> mergeBuildArgs(ImageConfiguration imageConfig, JKubeConfiguration configuration) {
+    return prepareBuildArgs(addBuildArgs(configuration), imageConfig.getBuildConfiguration());
+  }
```

Comment:
Could we use this PR to provide some sort of order and orchestration for this method instead of just chaining methods.

It'd be nice if the method would show how the args are computed and populated based on their source. This would also show how the precedence works.

```
final Map<String, String> args = new HashMap<>();
// Least precedence args added to the map
// Next args added to the map, overriding any previous entries
// ...
return args;
```

Reply:
While addressing this, I noticed that we're using [Guava's  `com.google.common.collect.ImmutableMap`](https://guava.dev/releases/23.0/api/docs/com/google/common/collect/ImmutableMap.html) that throws exception when colliding keys are added.

Shall I remove its usage? At the moment, we are throwing exception on collision.

Reply:
So at the moment it's not possible to override args? that's weird.
We should definitely remove Guava.
If the current behavior is that override is not possible, we can preserve it. We should also create a new issue to see if we want to change this. IMO users my want to be able to provide one-off overrides, so maybe a warning instead of an exception is better.

Reply:
I've created https://github.com/eclipse-jkube/jkube/issues/3000 for this task.